apps/sync/js/adapters/history.js

```diff
@@ -0,0 +1,283 @@
+  - Add a new or existed place to Places DataStore (PDS).
+    - Convert the records from HC to PDS format.
+    - Merge a new record and the existed one then write to PDS.
+  - Remove an existed place when receiving a "deleted: true" record.
+**/
+
+'use strict';
+
+/* global
+  asyncStorage,
+  DataAdapters,
+  LazyLoader
+*/
+
+const HISTORY_COLLECTION_MTIME = 'collections::history::mtime';
+const SYNCTOID_PREFIX = 'SynctoId::';
+
+var HistoryHelper = (() => {
+  var placesStore;
+  function _ensureStore() {
+    if (placesStore) {
+      return Promise.resolve(placesStore);
+    }
+    return navigator.getDataStores('places').then(stores => {
+      placesStore = stores[0];
+      return Promise.resolve(placesStore);
+    });
+  }
+
+  function setSyncedCollectionMtime(mtime) {
+    return new Promise(resolve => {
+      LazyLoader.load(['shared/js/async_storage.js'], () => {
+        asyncStorage.setItem(HISTORY_COLLECTION_MTIME, mtime, resolve);
+      });
+    });
+  }
+
+  function getSyncedCollectionMtime() {
+    return new Promise(resolve => {
+      LazyLoader.load(['shared/js/async_storage.js'], () => {
+        asyncStorage.getItem(HISTORY_COLLECTION_MTIME, resolve);
+      });
+    });
+  }
+
+  function setSynctoIdMap(synctoId, dsId) {
+    return new Promise(resolve => {
+      LazyLoader.load(['shared/js/async_storage.js'], () => {
+        asyncStorage.setItem(SYNCTOID_PREFIX + synctoId, dsId, resolve);
+      });
+    });
+  }
+
+  function getSynctoIdMap(synctoId) {
+    return new Promise(resolve => {
+      LazyLoader.load(['shared/js/async_storage.js'], () => {
+        asyncStorage.getItem(SYNCTOID_PREFIX + synctoId, resolve);
+      });
+    });
+  }
+
+  function mergeRecordsToDataStore(existed, newPlace) {
+    if (existed.url !== newPlace.url) {
+      // The existed record has different url(id) with the new record.
+      throw new Error('Inconsistent records');
+    }
+    if (!existed.fxsyncId) {
+      existed.fxsyncId = newPlace.fxsyncId;
+    } else if(existed.fxsyncId !== newPlace.fxsyncId) {
+      // Two records have different fxsyncId but have the same url(id).
+      throw new Error('Inconsistent records');
+    }
+
+    existed.visits = existed.visits || [];
+    if (existed.visits.length === 0 && newPlace.title) {
+      existed.title = newPlace.title;
+    } else if (newPlace.visits[0] >= existed.visits[0]) {
+      existed.title = newPlace.title;
+    }
+
+    newPlace.visits.forEach(item => {
+      if (existed.visits.indexOf(item) === -1) {
+        existed.visits.push(item);
+      }
+    });
+
+    existed.visits.sort((a, b) => {
+      // sort in descending order
+      return b - a;
+    });
+
+    return existed;
+  }
+
+  function addPlace(place) {
+    // 1. get place by url(id of DataStore)
+    // 2.A merge the existed one and new one if it's an existed one,
+    //     and update the places.
+    // 2.B Add a new record with RevisionId.
```

Comment:
... and add the DataStore record ID into the localID <-> remoteID matching table.